src/runtime/runtime_data.cs

```diff
@@ -0,0 +1,438 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Runtime.Serialization;
+using System.Runtime.Serialization.Formatters.Binary;
+
+using static Python.Runtime.Runtime;
+
+namespace Python.Runtime
+{
+    public static class RuntimeData
+    {
+        private static Type _formatterType;
+        public static Type FormatterType
+        {
+            get => _formatterType;
+            set
+            {
+                if (!typeof(IFormatter).IsAssignableFrom(value))
+                {
+                    throw new ArgumentException("Not a type implemented IFormatter");
+                }
+                _formatterType = value;
+            }
+        }
+
+        public static ICLRObjectStorer WrappersStorer { get; set; }
+
+        internal static void Stash()
+        {
+            var metaStorage = new RuntimeDataStorage();
+            MetaType.SaveRuntimeData(metaStorage);
+
+            var importStorage = new RuntimeDataStorage();
+            ImportHook.SaveRuntimeData(importStorage);
+
+            var typeStorage = new RuntimeDataStorage();
+            TypeManager.SaveRuntimeData(typeStorage);
+
+            var clsStorage = new RuntimeDataStorage();
+            ClassManager.SaveRuntimeData(clsStorage);
+
+            var moduleStorage = new RuntimeDataStorage();
+            SaveRuntimeDataModules(moduleStorage);
+
+            var objStorage = new RuntimeDataStorage();
+            SaveRuntimeDataObjects(objStorage);
+
+            var runtimeStorage = new RuntimeDataStorage();
+            runtimeStorage.AddValue("meta", metaStorage);
+            runtimeStorage.AddValue("import", importStorage);
+            runtimeStorage.AddValue("types", typeStorage);
+            runtimeStorage.AddValue("classes", clsStorage);
+            runtimeStorage.AddValue("modules", moduleStorage);
+            runtimeStorage.AddValue("objs", objStorage);
+
+            IFormatter formatter = CreateFormatter();
+            var ms = new MemoryStream();
+            formatter.Serialize(ms, runtimeStorage);
+
+            Debug.Assert(ms.Length <= int.MaxValue);
+            byte[] data = ms.GetBuffer();
+            // TODO: use buffer api instead
+            IntPtr mem = PyMem_Malloc(ms.Length + IntPtr.Size);
+            Marshal.WriteIntPtr(mem, (IntPtr)ms.Length);
+            Marshal.Copy(data, 0, mem + IntPtr.Size, (int)ms.Length);
+
+            IntPtr capsule = PySys_GetObject("clr_data").DangerousGetAddress();
```

Comment:
Why can't this work with `BorrowedReference` directly?

Reply:
It could, but it would not improve readability. While we do get a borrowed reference, we really do own this reference.

Reply:
I am not so sure. It took some time to figure out, that in 73-80 `capsule` is borrowed, in 80-82 it is a new object, and that in 81 the reference to it is not being stolen, and required decref in 82.

Reply:
Right, I get what you are saying now. I'll factor out the clearing.